File: 플젝1b/source/src/Assembler.java
import java.io.*;
import java.util.*;

public class Assembler {
	InstTable instTable;
	ArrayList<String> lineList;
	ArrayList<SymbolTable> symtabList;
	ArrayList<TokenTable> TokenList;
	ArrayList<LitTable> litList; 	//리터럴테이블
	ArrayList<String> codeList;
	ArrayList<ExtTable> ExtList; 	//외부주소 테이블
	
	public Assembler(String instFile) {
		instTable = new InstTable(instFile);
		lineList = new ArrayList<String>();
		symtabList = new ArrayList<SymbolTable>();
		TokenList = new ArrayList<TokenTable>();
		litList = new ArrayList<LitTable>();
		codeList = new ArrayList<String>();
		ExtList = new ArrayList<ExtTable>();
	}

	public static void main(String[] args) {
		
		Assembler assembler = new Assembler("inst.data");
		assembler.loadInputFile("input.txt");
		
		assembler.pass1();
		
		try{assembler.printSymbolTable("symtab_20142249.txt");}
		catch(IOException e) {
			System.out.println("출력중 오류 발생");
		}
		assembler.pass2();
		try{assembler.printObjectCode("output_20142249.txt");}
		catch(IOException e) {
			System.out.println("출력중 오류 발생");
		}
		
	}

	//오브젝트 코드를 출력한다.
	private void printObjectCode(String fileName) throws IOException {
		PrintWriter output = new PrintWriter(fileName);
		for(int i=0;i<TokenList.size();i++) {
			output.println(TokenList.get(i).FinalCode);
		}
		output.close();
	}
	
	//심볼테이블 출력부
	private void printSymbolTable(String fileName) throws IOException {

		PrintWriter output = new PrintWriter(fileName);
		for(int i =0;i<symtabList.size();i++) {
			for(int j = 0; j<symtabList.get(i).symbolList.size(); j++) {
				output.println(symtabList.get(i).symbolList.get(j));
				output.println(Integer.toHexString(symtabList.get(i).locationList.get(j)));
			}
		}
		output.close();
	}

	//토큰테이블, 심볼테이블, 리터럴 테이블을 생성하는 pass1()
	private void pass1() {
		int i = 0;
		int j = 0;
		
		//라인을 토큰으로 1:1 할당하는 과정이다.
		TokenTable temp= new TokenTable();
		TokenList.add(temp);
		for(;i<lineList.size();i++) {
			try{TokenList.get(j).putToken(lineList.get(i),instTable);}
			
			//섹터가 바뀔 경우 섹터체인지를 쓰로우하여 다음 리스트에 저장한다.
			catch(SectorChange e) {
				Token tempToken = TokenList.get(j).popToken();
				j++; 
				TokenList.add(new TokenTable());
				TokenList.get(j).pushToken(tempToken);
				continue;
			}
		}
		i=0;j=0;
		
		//리터럴 테이블을 생성하는 과정
		for(i=0; i<TokenList.size();i++) {
			try{
				litList.add(new LitTable());
				TokenList.get(i).linkLit(litList.get(i));
				TokenList.get(i).linkInst(instTable);
				TokenList.get(i).makeLit();}
			catch(SectorChange e) {
				continue;
				}
		}
		
		//생성된 리터럴 테이블에 따라 주소를 수정하는 과정
		for(i=0;i<TokenList.size();i++) {
			TokenList.get(i).calLit();
		}

		
		// 심볼 테이블을 생성하는 과정
		for(i=0;i<TokenList.size();i++) {
			try{
				symtabList.add(new SymbolTable());
				TokenList.get(i).linkSym(symtabList.get(i));
				TokenList.get(i).makeSymtab();}
			catch(SectorChange e) {
				continue;
			}
		}
		j=0;
		
		int t=5;
		// TODO Auto-generated method stub
		
	}
	

	/**
	 * pass2 과정을 수행한다.<br>
	 *   1) 분석된 내용을 바탕으로 object code를 생성하여 codeList에 저장.
	 */
	private void pass2() {
		
		//외부주소 테이블을 만드는 과정
		for(int i=0;i<TokenList.size();i++) {
			try{
				ExtList.add(new ExtTable());
				TokenList.get(i).linkExt(ExtList.get(i));
				TokenList.get(i).makeExt();
			}
			catch(SectorChange e) {
				continue;
			}
		}
		
		//기계어를 생성하는 과정
		for(int i=0;i<TokenList.size();i++) {
			try{
				TokenList.get(i).makeObjectCode();
			}
			catch(SectorChange e) {
				continue;
			}
		}
		int a;
		
		//출력물을 생성하는 과정
		for(int i=0;i<TokenList.size();i++) {
			try{
				TokenList.get(i).buildObjectCode();
			}
			catch(SectorChange e) {
				continue;
			}
		}
	}
	
	//inputFile을 읽어 라인별로 저장한다.
	private void loadInputFile(String inputFile) {
		try {
		File input = new File(inputFile);
		Scanner temp;
		temp = new Scanner(input);
		while(temp.hasNextLine()) {
			String line = temp.nextLine();
			if(line.startsWith(".")) continue;
			lineList.add(line);
		}
		temp.close();
	}
		catch(FileNotFoundException e){
			System.out.println("file not found");
			e.printStackTrace();
		}
	}
}

File: 플젝1c/source/Assembler_python/Assembler.py
#Assembler.py
import re,SectorChange,LitTable,ExtTable,SymbolTable,InstTable,TokenTable

class Assembler:
    def __init__(self):
        self.lineList=[ ]
        self.symtabList =[ ]
        self.TokenList =[ ]
        self.litList = [ ]
        self.codeList = [ ]
        self.ExtList =[ ]

        #명령어 테이블을 생성한다.
    def loadInstFile(self,inputFile):
        self.instTable = InstTable.InstTable(inputFile)

    def loadInputFile(self,inputFile):
        #라인을 입력받아 라인리스트에 저장
        try: 
            f = open(inputFile)
            while True:
              line = f.readline()
              if not line : break
              self.lineList.append(line)
        except:
            print('파일로드 실패')
        finally:
            f.close()
    
    def pass1(self):

        #라인에 따라 토큰테이블을 생성
        temp = TokenTable.TokenTable()
        for i in self.lineList:
            try:
                temp.linkInst(self.instTable)
                temp.putToken(i)
            except SectorChange.SectorChange: 
                tempToken = temp.popToken()
                self.TokenList.append(temp)
                temp = TokenTable.TokenTable() #수정 (push & pop)
                temp.pushToken(tempToken)
                continue
        self.TokenList.append(temp)

        #리터럴 테이블을 생성하고 리터럴 테이블의 주소를 계산
        for i in self.TokenList:
            try:
                temp = LitTable.LitTable()
                i.linkLit(temp)
                i.makeLit()
            except SectorChange.SectorChange:
                self.litList.append(temp)
                continue

        for i in self.TokenList:
            i.calLit()

        #심볼테이블을 생성하는 과정
        for i in self.TokenList:
            try:
                temp = SymbolTable.SymbolTable()
                i.linkSym(temp)
                i.makeSymtab()
            except SectorChange.SectorChange:
                self.symtabList.append(temp)
                continue

    #심볼테이블을 출력하는 과정
    def PrintSymbolTable(self,output):
        j=0
        f = open(output,'w')
        for i in self.symtabList:
            while j<len(i.symbolList):
                f.write(i.symbolList[j])
                f.write("\t")
                f.write(str(hex(i.locationList[j]))[2:])
                f.write("\n")
                j+=1
            f.write("\n")
            j=0
        f.close()

    def pass2(self):

        #외부주소 테이블을 만드는 과정
        for i in self.TokenList:
            try:
                temp = ExtTable.ExtTable()
                i.linkExt(temp)
                i.makeExt()
            except SectorChange.SectorChange:
                self.ExtList.append(temp)
                continue

        #기계어 코드를 생성하는 과정
        for i in self.TokenList:
            try:
                i.makeObjectCode()
            except SectorChange.SectorChange:
                continue

        #출력물을 생성하는 과정
        for i in self.TokenList:
            try:
                i.buildObjectCode()
            except SectorChange.SectorChange:
                continue

    def PrintObjectCode(self,output):
        j=0
        f = open(output,'w')
        for i in self.TokenList:
            f.write(i.FinalCode)
        f.close()